proposals/0106-paritioned-epoch-rewards-pda.md

```diff
@@ -0,0 +1,100 @@
+only way to find the stake or voting rewards for a specific address will be to
+iterate through blocks at the beginning of each epoch, inspecting all the
+rewards. This is because the runtime does not persist information about how the
+rewards were partitioned; in fact, it does not even persist how many blocks the
+rewards distribution spans, so there is no way to predict how long it will take
+or how far through the epoch to go in finding a particular address.
+
+## Alternatives Considered
+
+An alternative to storing partition data in an on-chain account would be to
+record the necessary data in the ledger in some fashion. This could be a
+transaction that gets added to the block, a special RewardType, or metadata that
+gets stored on the node and then duplicated to long-term storage, like Bigtable.
+In fact, it is probably worthwhile to pursue this alternative as well, since it
+will enable finding rewards without access to snapshots or the running chain.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+When partitioned rewards are calculated in the runtime (currently during the
+first block of the epoch), the runtime should populate a PDA that stores the
+partition data needed to recreate the hasher that returns the partition index
+for any address. This data comprises: the number of partitions, the parent
+blockhash, and the hasher kind. More specifically:
+
+```rust
+// Version wrapper to allow future updates
+// Variant serialized as little-endian unsigned 32-bit integer
+enum EpochRewardsPartitionDataVersion {
+    V0(PartitionData),
+}
+
+// Extensible list of kinds of hashers used to generate partitions
+// Mapping fn defines the HasherKind for each partition-data version
+enum HasherKind {
+    Sip13 = 0,
+}
+
+// Data about rewards partitions for a particular epoch
+struct PartitionData {
+    num_partitions: usize, // little-endian unsigned 64-bit integer
+    parent_blockhash: Hash, // byte-array of length 32
+}
+```
+
+The address of this PDA will use some bytes -- to prevent griefing and namespace
+the PDAs -- and the rewards distribution epoch number as a little-endian u64 as
+seeds. Specifically: `[b"EpochRewards",b"PartitionData", &epoch.to_le_bytes()]`.
+The owning program should be the sysvar program id:
+`Sysvar1111111111111111111111111111111111111`.
+
+## Impact
+
+The change in this proposal does increase the number of "forever" accounts that
+validators must store by one per epoch. However, the PDAs will be owned by the
+stake program, so could be adjusted or closed in the future by a feature-gated
+change to the Stake Program. Meanwhile, the change greatly improves the
+post-SIMD 0015 situation for clients trying to track stake or voting rewards,
+since they can use the data in the PDA to pull the correct partition directly,
+instead of scanning an unknown number of blocks.
+
+## Security Considerations
+
+Like traditional sysvars, the new accounts should only be loadable as read-only.
```

Comment:
The SIMD should explain how this is done since it's also important for consensus. Are they read-only only after they are created?

Reply:
As you stated on your SIMD, I think how dynamically addressed sysvars would have their write-locks demoted is outside the scope of this proposal. But I expanded the section a little.

Reply:
This SIMD needs to address the specifics of write-lock demotion if you state that these sysvars should be loaded as read-only. Right now we only have a way to use feature gates to introduce new read-only sysvars.. we don't have a design for this new kind of sysvar which is created each epoch yet. So I think this SIMD can reference SIMD 105 and say that the new sysvars will be added to the dynamic list when they are created on-chain. Also, I don't think that the "Security considerations" is the right place to discuss read-only loading.. it should be in the design section.

Reply:
I'm fine with handling in a separate SIMD but this SIMD says they should only be loadable as read-only but doesn't explain that the design is still TBD. So I think we should say that it's TBD or design it here.

Reply:
As for lack of persistence, we can add all these sysvars in a loop from some starting epoch to the current epoch. This seems fine to do in each restart because epochs will only be in the hundreds to thousands.

Reply:
I feel like the use of "should" here is consistent with rfc2119, and saying "a new proposal should be introduced" is the same thing as saying the design is TBD. But I'm also happy to remove the paragraph entirely.

Reply:
Oh sorry I didn't realize "should" meant that